### PR TITLE
ENH: Added analytical formula for dgamma distribution entropy (#17748)

### PR DESCRIPTION
**Pull request by MatteoRaso**

### CODE REVIEW

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:1520]: Compared to gh-17868, this one is more robust because you're already working with the log of the terms.

However, you'll still run into the second problem here: the `loggamma(a)` term and `a * digamma(a)` term will be huge and of opposite sign. There's a relationship between `loggamma(a)` and `a * digamma(a)` that you can use to cancel out most of them analytically, at least for large `a`. (Taking a look at their series expansions might help to find it.) 

In any case, this is why I suggested in gh-17211 that you open only one of these at a time. This is the same as one of the problems you'll encounter in gh-17868. If you fix the issues in one PR before opening the next, there will be fewer cycles of review in the following PRs.

**Comment by dschmitz89** [comment]
What is the reference for this formula? Should be added as a comment in the code or at least be mentioned here for future maintenance.

**Comment by MatteoRaso** [comment]
> What is the reference for this formula? Should be added as a comment in the code or at least be mentioned here for future maintenance.

I had to derive the formula myself. Sadly, there doesn't seem to many good references for any entropy formulas, but it's not terribly hard to verify that the formula is correct.

File: scipy/stats/_continuous_distns.py
Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py:1520]: I tested dgamma with ```a = 1e100``` and ```a = 1e-100```, and there was no over/underflow. Unless I'm missing something, the formula should be fine as is.

Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:1520]:
```python3
a = 1
b = np.pi * 1e-17
c = -1
print(a + b + c)
```

**Comment by dschmitz89** [comment]
> > What is the reference for this formula? Should be added as a comment in the code or at least be mentioned here for future maintenance.
> 
> I had to derive the formula myself. Sadly, there doesn't seem to many good references for any entropy formulas, but it's not terribly hard to verify that the formula is correct.

The test against the numerical integration seems to disagree though at the moment: 
https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=22918&view=logs&j=7902a2c1-7b88-5b2d-6408-84b88759c642&t=5471a576-12f4-5533-4de4-506800a17112&l=852

image

These tests run only with additional arguments. To test locally, do 
```
export SCIPY_XSLOW=1
python dev.py -t [yourtest] -m full
```

**Comment by MatteoRaso** [comment]
Good catch, @dschmitz89. I looked it over, and found that I incorrectly integrated from 0 to infinity, when I should have integrated from -infinity to infinity. Error has been fixed.

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [reply scipy/stats/_continuous_distns.py:1520]: Even if there is no over/underflow, the result can become imprecise (see the post above). Therefore tests with small values and big values should be added ( e.g. `m=1e-4` and `m=1e20`)

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3440]:
```suggestion
        assert_allclose(stats.dgamma.entropy(a), ref)
```

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3444]:
```suggestion
```

These should be removed in favor of actual tests against the result computed with mpmath for extreme values for `m`  (see above).

**Review by dschmitz89 — COMMENTED** [review]
Regarding lint errors: at the moment, scipy`s linting process is undergoing a major refactoring. Let's worry about that later.

**Comment by dschmitz89** [comment]
> > What is the reference for this formula? Should be added as a comment in the code or at least be mentioned here for future maintenance.
> 
> I had to derive the formula myself. Sadly, there doesn't seem to many good references for any entropy formulas, but it's not terribly hard to verify that the formula is correct.

Could you do provide a rough derivation how you solved that integral here (github markdown supports LateX for a few months now) or upload a PDF with it?

**Comment by MatteoRaso** [comment]
> > > What is the reference for this formula? Should be added as a comment in the code or at least be mentioned here for future maintenance.
> > 
> > 
> > I had to derive the formula myself. Sadly, there doesn't seem to many good references for any entropy formulas, but it's not terribly hard to verify that the formula is correct.
> 
> Could you do provide a rough derivation how you solved that integral here (github markdown supports LateX for a few months now) or upload a PDF with it?
[Dgamma_entropy_derivation.pdf](https://github.com/scipy/scipy/files/10541192/Dgamma_entropy_derivation.pdf)

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:1520]:
```suggestion
        h = a + np.log(2) + sc.gammaln(a) + (1 - a) * sc.digamma(a)
```

Please prefer `gammaln` over `loggamma`. `loggamma` is not used for any other real distribution and slower than `gammaln`:

```
%timeit -n 100000 sc.loggamma(1) # 697 ns ± 34.9 ns per loop
%timeit -n 100000 sc.gammaln(1) # 670 ns ± 13.5 ns per loop
```

**Comment by dschmitz89** [comment]
This implementation is definitely an improvement in terms of speed and accuracy.

*Small a*

Dgamma_Entropy_small_a

<details>

import numpy as np
import matplotlib.pyplot as plt
from mpmath import mp
from scipy import stats
from scipy import special as sc

def entropy_dgamma(a):
    return a + np.log(2) + sc.gammaln(a) + (1 - a) * sc.digamma(a)

@np.vectorize
def entropy_dgamma_mpmath(a):
    a = mp.mpf(a)
    return a + mp.log(2) + mp.loggamma(a) + (1 - a) * mp.digamma(a)

x = np.logspace(5, 15, 100)
plt.semilogx(x, np.asarray(entropy_dgamma_mpmath(x), dtype=np.float64), label="mpmath", ls="dashed")
plt.semilogx(x, entropy_dgamma(x), label="custom", ls="dotted")
#plt.semilogx(x, stats.dgamma.entropy(x), label="main", ls="dashdot")
plt.legend()

</details>

*Large a*

Here, the main branch results in an error (already for ca. `a>100`) due to intermediate overflow. With this instead we get reasonable values untilt 1e12:

Dgamma_Entropy_large_a

I say, as for #17868 : let's add test cases for `1e-4` and `1e6` and then merge.

**Comment by MatteoRaso** [comment]
> let's add test cases for 1e-4 and 1e6 and then merge.

I prefer to add more test cases, since there can be failures in between these 2 values, even if both pass.

**Comment by dschmitz89** [comment]
Using the asymptotic formulas from https://github.com/scipy/scipy/pull/17868#issuecomment-1416919700 it is possible to get it right for $a>1e12$ (thanks @steppi !):

$$
\begin{align}
\ln(\Gamma(a)) & \sim (a -\frac{1}{2})\ln a - a + \frac{1}{2}\ln(2\pi) & \\\
\psi(a) & \sim \ln a - \frac{1}{2a}  \\\
h & = \ln 2 + \ln(\Gamma(a)) + a + (1 -a)\psi(a) \\\
& \sim \ln 2 + (a -\frac{1}{2})\ln a - a + \frac{1}{2}\ln(2\pi) + a + (1-a)(\ln a - \frac{1}{2a}) \\\
& = \ln 2 + \frac{1}{2}(1+\ln a + \ln(2\pi))
\end{align}
$$

Dgamma_Entropy_large_a

<details>

import numpy as np
import matplotlib.pyplot as plt
from mpmath import mp
from scipy import stats
from scipy import special as sc

def entropy_dgamma(a):
    return np.where(a < 1e12,
                    sc.gammaln(a) + a + np.log(2) + (1 - a) * sc.digamma(a), 
                    np.log(2) + 0.5 * (1. + np.log(a) + np.log(2*np.pi))
    )

@np.vectorize
def entropy_dgamma_mpmath(a):
    a = mp.mpf(a)
    return a + mp.log(2) + mp.loggamma(a) + (1 - a) * mp.digamma(a)

x = np.logspace(11, 14, 1000)
plt.semilogx(x, np.asarray(entropy_dgamma_mpmath(x), dtype=np.float64), label="mpmath", ls="dashed")
plt.semilogx(x, entropy_dgamma(x), label="custom", ls="solid")
plt.legend()
plt.show()

</details>

@MatteoRaso : would you prefer to leave this for a follow-up PR? I leave it up to you, does not have to hold up this PR.

**Comment by dschmitz89** [comment]
The `Dgamma` test class from in the main branch is not merged into this branch yet. When that is done, we can have a look at that last lint error. :)

**Comment by MatteoRaso** [comment]
>The Dgamma test class from in the main branch is not merged into this branch yet.

For some reason, when I try to merge main branch into my local copy of this branch, it causes a whole bunch of errors. Is it necessary to merge the main branch into this one, though? I could be wrong, but I feel like it should be possible to merge this branch without needing to worry about that.

>When that is done, we can have a look at that last lint error. :)

I ran ``python dev.py lint`` and ``python runtests.py --pep8``, but nothing came up for _continuous_distns.py or the test file. It's possible that the lint error is actually the linter being weird again.

**Comment by j-bowhay** [comment]
> I ran `python dev.py lint` and `python runtests.py --pep8`, but nothing came up for _continuous_distns.py or the test file. It's possible that the lint error is actually the linter being weird again.

There have been a lot of changes to the lint infrastructure lately; unless you have merged main in recently it is likely that the check you are running locally is different to what is running on the CI job.

**Comment by dschmitz89** [comment]
> > The Dgamma test class from in the main branch is not merged into this branch yet.
> 
> For some reason, when I try to merge main branch into my local copy of this branch, it causes a whole bunch of errors. Is it necessary to merge the main branch into this one, though? I could be wrong, but I feel like it should be possible to merge this branch without needing to worry about that.
> 
> > When that is done, we can have a look at that last lint error. :)
> 
> I ran `python dev.py lint` and `python runtests.py --pep8`, but nothing came up for _continuous_distns.py or the test file. It's possible that the lint error is actually the linter being weird again.

For the reasons @j-bowhay mentioned, merging main into this branch can become quite messy probably. I do not think we can come around that unfortunately though.

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:1621]: All of scipy's distributions support numpy arrays in their methods. We have to make sure that if a user provides a NumPy array of several values, the correct branch is chosen. Throughout this file, there is a convenience function called `_lazywhere` we should use here too.

The cutoff should also be lowered. For `1e13`, the regular formula becomes very imprecise already.

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:1625]:
```suggestion
```

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3881]:
```suggestion
```

Two tests with small `a` and two with large `a` should suffice.

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3891]:
```suggestion
```

File: scipy/stats/tests/test_distributions.py
Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py:3881]: >Two tests with small a and two with large a should suffice.

I'm not too sure about that. When I was working on the Nakagami entropy, I noticed that there were sometimes values where it would fail, even though values higher and lower than it would pass. It's not a huge deal to me, but I would prefer keeping these values.

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:3881]: In that case, maybe you should add values in between these?

Comment (dschmitz89) [reply scipy/stats/tests/test_distributions.py:3881]: Ok, that sounds reasonable. What about the following test cases for example: `1e-8`, `1e-2`, `1`, `1e2`, `1e4`, `1e8`, `1e13`

File: scipy/stats/_continuous_distns.py
Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py]: Done.

File: scipy/stats/tests/test_distributions.py
Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py]: Ok, I've added some more test values.

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3911]:
```suggestion
    [(1e-10, -9999999975.858217),
     (1e-5, -99987.3711165702),
     (1e1, 3.2292013590409248),
     (1e4, 6.717222565586032),
     (1e10, 13.625011178701513),
```

Comment (dschmitz89) [reply scipy/stats/tests/test_distributions.py:3911]: Feel free to add up to two more test cases but having more than 30 tests for such a small function  is just way too much. I do not think any other function of the univariate distributions is tested that extensively.

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:1615]: This might not be needed. Could you test if tests pass without this line? Other overriden distribution methods do not use it unless they use indexing.

Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py:1615]: No, a bunch of tests fail if that line is removed (44, to be exact).

File: scipy/stats/tests/test_distributions.py
Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py:3911]: Ok, I removed most of the tests and added 1e4, but we also need to keep the tests for 1e100 and 1e-100.

**Comment by MatteoRaso** [comment]
Ok, I've removed the pep8 errors that slipped under the linter and all tests pass. @dschmitz89 @mdhaber does this look good to merge?

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3892]:
```suggestion
        # def second_dgamma(a):
        #     a = mp.mpf(a)
        #     if a < 1e15:
        #         x_1 = a + mp.log(2) + mp.loggamma(a)
        #         x_2 = (mp.one - a) * mp.digamma(a)
        #         h = x_1 + x_2
        #
        #     else:
        #         x_1 = mp.one / mp.mpf(2) * (mp.one + mp.log(a) + mp.log(2 * mp.pi))
        #         h = mp.log(2) + x_1
```

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3869]:
```suggestion
        # The reference values were calculated with mpmath:
        # def entropy_dgamma(a):
        #    def pdf(x):
        #        A = mp.one / (mp.mpf(2.) * mp.gamma(a))
        #        B = mp.fabs(x) ** (a - mp.one)
        #        C = mp.exp(-mp.fabs(x))
        #        h = A * B * C
        #        return h
        #
        #    return -mp.quad(lambda t: pdf(t) * mp.log(pdf(t)),
```

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:1624]:
```suggestion
        h = _lazywhere(a > 1e10, (a), f=h2, f2=h1)
```

Turns out problems start way before `1e13`:

Dgamma_Entropy_large_a_problems

Zoomed in:

Dgamma_Entropy_large_a_problems_zoom

With a lower cutoff, we get around this.

**Review by dschmitz89 — COMMENTED** [review]
We have to be careful with `mpmath`. I made the same mistakes lately, resulting in inaccurate reference values. Please have a look at the suggestions. A general use of thumb could probably be to use mpmath functions wherever possible.

**Comment by mdhaber** [comment]
@dschmitz89 Please note that the bigger issue here is that `import mpmath as mp` needs to be `from mpmath import mp`. Please see https://github.com/scipy/scipy/pull/17874#discussion_r1105067620, mpmath/mpmath#657.

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3882]:
```suggestion
        # from mpmath import mp
```

**Comment by mdhaber** [comment]
Also, IIUC, the intent of using `mp.one` is to ensure that quantities that cannot be represented exactly in double precision are not truncated to double precision prematurely; i.e. `mp.one/3` or `mp.mpf('1/3')` would be preferred over `1/3`. It is a good thing to watch out for, but I don't think they make a difference here because the quantities they operate against are already `mpf`s.

@dschmitz89 I'll review/merge this when it looks good to you.

File: scipy/stats/tests/test_distributions.py
Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py]: This change didn't change the reference values when I reran it, so we don't need to worry about those.

Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py]: This change didn't change the reference values when I reran it, so we don't need to worry about those.

**Comment by MatteoRaso** [comment]
@dschmitz89 I've made the changes that you requested. Does this look ready to merge?

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3903]:
```suggestion
    def test_entropy_array_input(self):
        x = np.array([1, 5, 1e20, 1e-5])
        y = stats.dgamma.entropy(x)
        for i in range(len(y)):
              y[i] == stats.dgamma.entropy(x[i])
```

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3893]: With mpmath the asymptotic expansion is not needed, it maintains precision at high values if everything is *mpmathified*. I think I also messed that up in my [previous comment](https://github.com/scipy/scipy/pull/17873#issuecomment-1417065945) by using `1` and not `mp.one`, sorry about that. If we use the original formula instead, this also tests the correctness of the formula we obtained using the asymptotic expansion. What do you think? 

Using plain mpmath I get a nice smooth curve:


image

<details>

```
from mpmath import mp
import numpy as np
import matplotlib.pyplot as plt

mp.dps = 50

@np.vectorize
def second_dgamma(a):
    a = mp.mpf(a)
    x_1 = a + mp.log(2) + mp.loggamma(a)
    x_2 = (mp.one - a) * mp.digamma(a)
    h = x_1 + x_2
    return float(h)

x = np.logspace(0, 20, 1000)
plt.semilogx(x, np.array(second_dgamma(x), dtype=np.float64))
plt.show()

Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py:3893]: >this also tests the correctness of the formula we obtained using the asymptotic expansion

I doubt that this is necessary. The asymptotic expansion was derived analytically, so it shouldn't need a test to confirm.

Comment (steppi) [reply scipy/stats/tests/test_distributions.py:3893]: > With mpmath the asymptotic expansion is not needed, it maintains precision at high values if everything is _mpmathified_. I think I also messed that up in my [previous comment](https://github.com/scipy/scipy/pull/17873#issuecomment-1417065945) by using `1` and not `mp.one`, sorry about that. If we use the original formula instead, this also tests the correctness of the formula we obtained using the asymptotic expansion. What do you think?

I agree we shouldn't use the asymptotic expansion in the mpmath baseline. 


> I doubt that this is necessary. The asymptotic expansion was derived analytically, so it shouldn't need a test to confirm.

It's still a good sanity check. On principle, I think it's best if the `mpmath` baseline defines the function as simply as possible directly from definitions which can be checked in a standard reference.

Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py:3893]: >With mpmath the asymptotic expansion is not needed, it maintains precision at high values if everything is mpmathified. 

I tried to generate the values without the asymptotic expansion, and it starts to run into problems for very large numbers. If we want to do a sanity check, we could just add a test value for some x where 1e10 < x < 1e15, the reason being that the formula uses the asymptotic expansion for x > 1e10 and the function that generates the reference values uses it for x > 1e15.

Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py]: Ok, done.

**Comment by dschmitz89** [comment]
I performed a more thorough error analysis following the discussion in #17930 . Let's change the threshold between regular formula and asymptotic formula to `1e8`. A relative error of `1e-9` is not good but should not hold up this PR in my opinion. Follow up could be to check if using more terms in the asymptotic expansion improve the result but I doubt it (did not work for the `chi` distribution).

image

<details>
<summary>Generate the above plot</summary>

```
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from scipy import special as sc
from mpmath import mp

mp.dps = 500

def dgamma_entropy(a):
    return a + sc.gammaln(a) + np.log(2) + (1 - a) * sc.digamma(a)

def asymptotic(a):
    return np.log(2) + 0.5 * (1 + np.log(a) + np.log(2 * np.pi))

@np.vectorize
def dgamma_entropy_mpmath(a):
    a = mp.mpf(a)
    two = mp.mpf(2.)
    x_1 = a + mp.log(two) + mp.loggamma(a)
    x_2 = (mp.one - a) * mp.digamma(a)
    h = x_1 + x_2
    return float(h)

x = np.logspace(2, 15, 1000)
reference = np.array(dgamma_entropy_mpmath(x), np.float64)
asymptotic_res = asymptotic(x)
regular_res = dgamma_entropy(x)

plt.loglog(x, np.abs((regular_res - reference)/reference), label="regular", ls='dashed')
plt.loglog(x, np.abs((asymptotic_res - reference)/reference), label="asymptotic", ls='dotted')
plt.axvline(1e8, 0, 1, c='k', label="Threshold")
plt.title("Relative error: dgamma entropy")
plt.legend()
plt.show()

```

</details>

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:1624]:
```suggestion
        h = _lazywhere(a > 1e8, (a), f=h2, f2=h1)
```

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
Last necessary change from my side. The lint error stems from a line not touched in this PR, might be fixed by merging main into this branch again.

**Review by dschmitz89 — APPROVED** [review]

**Comment by mdhaber** [comment]
@steppi thanks for your review. Feel free to merge when it looks good to you.

**Pull request merged**
